universal-application-tool-0.0.1/app/auth/SamlCiviFormProfileAdapter.java

```diff
@@ -0,0 +1,172 @@
+      return Optional.empty();
+    }
+
+    SAML2Profile profile = (SAML2Profile) samlProfile.get();
+
+    // Check if we already have a profile in the database for the user returned to us by SAML2
+    Optional<Applicant> existingApplicant =
+        applicantRepositoryProvider
+            .get()
+            .lookupApplicant(profile.getAttribute("email", String.class))
+            .toCompletableFuture()
+            .join();
+
+    // Now we have a three-way merge situation.  We might have
+    // 1) an applicant in the database (`existingApplicant`),
+    // 2) a guest profile in the browser cookie (`existingProfile`)
+    // 3) a SAML2Profile in the callback from the Identity Provider (`profile`).
+    // We will merge 1 and 2, if present, into `existingProfile`, then merge in `profile`.
+
+    Optional<CiviFormProfile> existingProfile = profileUtils.currentUserProfile(context);
+    if (existingApplicant.isPresent()) {
+      if (existingProfile.isEmpty()) {
+        // Easy merge case - we have an existing applicant, but no guest profile.
+        // This will be the most common.
+        existingProfile = Optional.of(profileFactory.wrap(existingApplicant.get()));
+      } else {
+        // Merge the two applicants and prefer the newer one.
+        // For account, use the existing account and ignore the guest account.
+        Applicant guestApplicant = existingProfile.get().getApplicant().join();
+        Account existingAccount = existingApplicant.get().getAccount();
+        Applicant mergedApplicant =
+            applicantRepositoryProvider
+                .get()
+                .mergeApplicants(guestApplicant, existingApplicant.get(), existingAccount)
+                .toCompletableFuture()
+                .join();
+        existingProfile = Optional.of(profileFactory.wrap(mergedApplicant));
+      }
+    }
+
+    // Now merge in the information sent to us by the SAML Identity Provider.
+    if (existingProfile.isEmpty()) {
+      LOG.debug("Found no existing profile in session cookie.");
+      return Optional.of(civiformProfileFromSamlProfile(profile));
+    } else {
+      return Optional.of(mergeCiviFormProfile(existingProfile.get(), profile));
+    }
+  }
+
+  public CiviFormProfileData civiformProfileFromSamlProfile(SAML2Profile profile) {
+    return mergeCiviFormProfile(
+        profileFactory.wrapProfileData(profileFactory.createNewApplicant()), profile);
+  }
+
+  public CiviFormProfileData mergeCiviFormProfile(
+      CiviFormProfile civiFormProfile, SAML2Profile saml2Profile) {
+    final String locale = saml2Profile.getAttribute("locale", String.class);
+    final boolean hasLocale = !Strings.isNullOrEmpty(locale);
+
+    final String firstName = saml2Profile.getAttribute("first_name", String.class);
+    final boolean hasFirstName = !Strings.isNullOrEmpty(firstName);
+
+    // TODO: figure out why the last_name attribute is being returned as an ArrayList because this
```

Comment:
Is it possible that all attributes can be repeated?

Reply:
Not sure. This is the only attribute that's being returned as a list, and it wasn't configured any differently from the others. FirstName is still a string.